### ISSUE TITLE
SQLite crashes triggered from Airship maintenance tasks

### ISSUE THREAD

**Issue by darrentaft** [urbanairship/android-library#165]
❗For how-to inquiries involving Airship functionality or use cases, please
contact (support)[https://support.airship.com/].

# Preliminary Info

### What Airship dependencies are you using?
urbanairship-fcm:14.0.2
urbanairship-automation:14.0.2
urbanairship-message-center:14.0.2

### What are the versions of any relevant development tools you are using?
N/A

# Report

### What unexpected behavior are you seeing?
Crashes from Airship SDK, reported to Google Play.

### What is the expected behavior?
Don't crash

### What are the steps to reproduce the unexpected behavior?
Unknown

### Do you have logging for the issue?

2 crashes reported to Google Play in the SQLite part of Airship - each reported many times:

Reported 467 times from 2 users:

```
android.database.sqlite.SQLiteException: 
  at android.database.sqlite.SQLiteConnection.nativeOpen (SQLiteConnection.java)
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:210)
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:194)
  at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked (SQLiteConnectionPool.java:493)
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:200)
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:192)
  at android.database.sqlite.SQLiteDatabase.openInner (SQLiteDatabase.java:864)
  at android.database.sqlite.SQLiteDatabase.open (SQLiteDatabase.java:849)
  at android.database.sqlite.SQLiteDatabase.openDatabase (SQLiteDatabase.java:724)
  at android.database.sqlite.SQLiteDatabase.openDatabase (SQLiteDatabase.java:714)
  at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked (SQLiteOpenHelper.java:295)
  at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase (SQLiteOpenHelper.java:238)
  at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase (FrameworkSQLiteOpenHelper.java:92)
  at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase (FrameworkSQLiteOpenHelper.java:53)
  at androidx.room.RoomDatabase.inTransaction (RoomDatabase.java:476)
  at androidx.room.RoomDatabase.assertNotSuspendingTransaction (RoomDatabase.java:281)
  at com.urbanairship.automation.storage.AutomationDao_Impl.getActiveExpiredSchedules (AutomationDao_Impl.java:1420)
  at com.urbanairship.automation.AutomationEngine.cleanSchedules (AutomationEngine.java:892)
  at com.urbanairship.automation.AutomationEngine.access$600 (AutomationEngine.java:72)
  at com.urbanairship.automation.AutomationEngine$4.run (AutomationEngine.java:242)
```

Reported 333 times from 4 users:

```
android.database.sqlite.SQLiteCantOpenDatabaseException: 
  at android.database.sqlite.SQLiteConnection.nativeOpen (SQLiteConnection.java)
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:300)
  at android.database.sqlite.SQLiteConnection.open (SQLiteConnection.java:218)
  at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked (SQLiteConnectionPool.java:737)
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:284)
  at android.database.sqlite.SQLiteConnectionPool.open (SQLiteConnectionPool.java:251)
  at android.database.sqlite.SQLiteDatabase.openInner (SQLiteDatabase.java:1394)
  at android.database.sqlite.SQLiteDatabase.open (SQLiteDatabase.java:1339)
  at android.database.sqlite.SQLiteDatabase.openDatabase (SQLiteDatabase.java:974)
  at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked (SQLiteOpenHelper.java:448)
  at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase (SQLiteOpenHelper.java:391)
  at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper$OpenHelper.getWritableSupportDatabase (FrameworkSQLiteOpenHelper.java:92)
  at androidx.sqlite.db.framework.FrameworkSQLiteOpenHelper.getWritableDatabase (FrameworkSQLiteOpenHelper.java:53)
  at androidx.room.RoomDatabase.inTransaction (RoomDatabase.java:476)
  at androidx.room.RoomDatabase.assertNotSuspendingTransaction (RoomDatabase.java:281)
  at com.urbanairship.automation.storage.AutomationDao_Impl.getActiveExpiredSchedules (AutomationDao_Impl.java:1420)
  at com.urbanairship.automation.AutomationEngine.cleanSchedules (AutomationEngine.java:892)
  at com.urbanairship.automation.AutomationEngine.access$600 (AutomationEngine.java:72)
  at com.urbanairship.automation.AutomationEngine$4.run (AutomationEngine.java:242)
```

**Comment by rlepinski** [comment]
Thanks for the report, we will take a look.

**Comment by rlepinski** [comment]
Do you have any updates on # of users affected? Could you post any other device information like manufacturer? 

I am pretty sure it has to do with these handful of users just not having enough space on the device, but would like to know if its a bigger issue before I just blindly try/catch all sql exceptions.

**Comment by darrentaft** [comment]
14 users, 5.46k crashes so far.

The following devices listed:

- Motorola moto e5
- Google AMD Stoney Ridge Chromebook
- Acer Iconia Tab 10
- Motorola moto e5 play
- Nokia 1.3
- Google Chromebook 14 (CB3-431)
- Samsung Galaxy Tab S5e
- Fluo F2
- Motorola moto e5 play
- Alba Alba8Nou

Thanks

**Comment by rlepinski** [comment]
Do you have an idea of your total audience size? 14 users is not that many, maybe just storage space issue?

**Comment by darrentaft** [comment]
570k unique users in the last 30 days, so yes - the number is small, but it's putting a dent in my stability stats.  ;-)

**Comment by rlepinski** [comment]
I have a fix in review, it just catches the exceptions and no-ops on the operation. It is what we were doing pre-room migration so I believe it should be fine. Should be out today or tomorrow.

**Comment by darrentaft** [comment]
Great. Thanks Ryan.

**Comment by rlepinski** [comment]
Fixed in 14.0.3
